optimizer/lib/transformers/BlurImagePlaceholders.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+var sizeOf = require('image-size');
+var jimp = require('jimp');
+
+/**
+ * Adds placeholders for all AMP images that are blurry versions of the
+ * corresponding original source. The blur will be displayed as the <amp-img> is
+ * loading, and
+ */
+
+class BlurImagePlaceholders {
+
+  transform(tree, params) {
+    const html = tree.root.firstChildByTag('html');
+    const body = html.firstChildByTag('body');
+    let node = body;
+    var ampImgPromises = [];
+    while (node !== null) {
+      if (node.tagName === 'amp-img') {
+        let parent = node;
+        ampImgPromises.push(this.addBitmap(tree, parent).then
+        (imgChild => {
+            console.log("4");
+            parent.appendChild(imgChild);
+        }));
+      }
+        node = node.nextNode();
+    }
+    return Promise.all(ampImgPromises);
+  }
+
+  nextNode(node) {
+    if (node.nextSibling) {
+      return node.nextSibling;
+    }
+    return this.nextNode(node.parent);
+  }
+
+  addBitmap(tree, node) {
```

Comment:
We should check for an existing blurry placeholder and early-out if there is one.